### PR TITLE
Bring in changes to upload scripts from PX4/Firmware

### PR DESCRIPTION
**Pull request by lucasdemarchi**

### CODE REVIEW

**Comment by lucasdemarchi** [comment]
This fixes uploading to aerofc with multiple baud rates.

**Comment by OXINARF** [comment]
The commit `Uploader: Signal GCS or other connected devices that a FiFirmware upload is pending and requires device access.` looks very arbitrary (besides unreadable code, what command is it sending?): so you want an eventual GCS to free the serial port so it sends an UDP MAVLink packet to some port in localhost?

Also the commit `Uploader: Do not show timeout for larger binaries when everything is still as expected.` should probably correct the comment on line 350.

Other than this, LGTM.

**Comment by lucasdemarchi** [comment]
> The commit Uploader: Signal GCS or other connected devices that a FiFirmware upload is pending and requires device access. looks very arbitrary (besides unreadable code, what command is it sending?):

This is bringing the commits from "upstream" location, where the script has been done. So if it doesn't do any harm, I'd prefer to bring them as is to make it easier in future to keep them synchronized.

> so you want an eventual GCS to free the serial port so it sends an UDP MAVLink packet to some port in localhost?

This is sending to a locally-connected GCS (127.0.0.1:14550) to free up the serial port. I agree the way it's done is not optimal, but I still prefer to bring it as is since it's harmless.

> 
> Also the commit Uploader: Do not show timeout for larger binaries when everything is still as expected. should probably correct the comment on line 350.

I can do that on top since the commit was not mine.

File: Tools/px_uploader.py [hunk 1]
```diff
@@ -620,6 +625,27 @@ def main():
...
+    # tell any GCS that might be connected to the autopilot to give up
+    # control of the serial port
+
+    # send to localhost and default GCS port
+    ipaddr = '127.0.0.1'
+    portnum = 14550
+
+    # COMMAND_LONG in MAVLink 1
+    heartbeatpacket = bytearray.fromhex('fe097001010000000100020c5103033c8a')
+    commandpacket = bytearray.fromhex('fe210101014c00000000000000000000000000000000000000000000803f00000000f6000000008459')
```
Comment (tridge) [inline Tools/px_uploader.py]: what is this packet?

**Comment by tridge** [comment]
> This is sending to a locally-connected GCS (127.0.0.1:14550) to free up the serial port. I agree the way it's done is not optimal, but I still prefer to bring it as is since it's harmless.

I think we need to remove or revert that commit, or add a command line option to enable it. If I'm running a SITL test in one window and upload a firmware to a pixhawk in another then I don't expect the upload of a firmware to a serial port to kill my SITL session. The serial port and the UDP link are completely separate.
I can see why it would be useful for the special case of something like the aero-fc, but it is a bad idea as a unconditional feature.

**Comment by tridge** [comment]
i think we should consider having our own uploader.py script in Tools/scripts in ArduPilot. We're going to need changes for new board types and for new USB IDs.

**Comment by lucasdemarchi** [comment]
> I think we need to remove or revert that commit, or add a command line option to enable it. If I'm running a SITL test in one window and upload a firmware to a pixhawk in another then I don't expect the upload of a firmware to a serial port to kill my SITL session. The serial port and the UDP link are completely separate.

That's a good reason not to bring the commit in. I'll remove it.

**Comment by lucasdemarchi** [comment]
> i think we should consider having our own uploader.py script in Tools/scripts in ArduPilot. We're going to need changes for new board types and for new USB IDs.

It's already different. Changes the other way around are not being applied.

**Comment by lucasdemarchi** [comment]
Requested changes done.

**Review by OXINARF — APPROVED** [review]
LGTM

**Pull request merged**